src/engineer/nowcast.py

```diff
@@ -0,0 +1,124 @@
+import numpy as np
+import calendar
+from datetime import date
+from pathlib import Path
+import xarray as xr
+import warnings
+
+from typing import cast, Dict, Optional, Tuple
+
+from ..utils import minus_months
+from .engineer import Engineer
+
+
+class NowcastEngineer(Engineer):
+    """Engineer the preprocessed `.nc` files into `/train`, `/test` `{x, y}.nc`
+    for the `nowcast` experiment.
+
+    This takes all non_target variables UP TO the target timestep and the
+    target variables for all previous timesteps (not the target timestep).
+
+    This produces a dataset:
+        t-3  t-2  t-1  t=0 |  y
+        ------------------------
+        P V  P V  P V  P   |  V
+                           |  V
+
+    Where P is the `non_target_variable` and V is the `target_variable`.
+    t=0 is the `target_timestep`.
+    """
+
+    def __init__(self, data_folder: Path = Path("data")) -> None:
+        self.name: str = "nowcast"
+
+        super().__init__(self.name, data_folder)
+
+    def stratify_xy(
+        self,
+        ds: xr.Dataset,
+        year: int,
+        target_variable: str,
+        target_month: int,
+        pred_months: int,
+        expected_length: Optional[int] = None,
+    ) -> Tuple[Optional[Dict[str, xr.Dataset]], date]:
+        """
+        The nowcasting experiment has different lengths for the
+         `target` variable vs. the `non_target` variables.
+
+        e.g. if I set the `pred_months = 11`
+
+        `x_target_variable` = 11 timesteps
+        `x_non_target_variable` = 12 timesteps (`pred_months + 1`).
+
+        We overcome this by creating an extra timestep with all nan values
+        in the `x_dataset`. This way the `x_dataset` contains the `y_dataset`
+        timestep but the `target_variable` is an array of all `np.nan` for that
+        target timestep. This prevents model leakage.
+        """
+        if expected_length is not None:
+            warnings.warn("Expected length must be None\
+                          for the `nowcast` experiment. Forcing to\
+                          None")
+            expected_length = None
```

Comment:
As we discussed, the expected length check got triggered quite often. Is there a way to consider it anyway?

One option would be to have it for the `non_target_dataset`, and then `expected_length + 1` for the `target_dataset`

Reply:
okay so i removed this line and added the following below:

```
        if expected_length is not None:
            # filter for missing values in timesteps!
            if sum(x_target_dataset) != expected_length:
                print(f"Wrong number of x values! Got {sum(x)} Returning None")

                return None, cast(date, max_train_date)

            if sum(x_non_target_dataset) != expected_length + 1:
                print(f"Wrong number of x values! Got {sum(x)} Returning None")

                return None, cast(date, max_train_date)

```

Reply:
oaky added a warning if the value is `None` too

Reply:
Thanks Tommy!